### PR TITLE
explicitly close source stream on local / encryption storage

### LINKED ISSUE #13276: Can't upload empty file when encryption app is enabled
### Steps to reproduce
1. create empty file (e.g. `touch emptyfile`)
2. upload the empty file using web or client

### Expected behaviour
The file should be uploaded fine.

### Actual behaviour
The file isn't uploaded (error 500) and an error is reported.

### Server configuration

**Operating system**: Centos 7.6

**Web server:** Apache/2.4.6

**Database:** mysql  Ver 15.1 Distrib 5.5.60-MariaDB,

**PHP version:** PHP 7.3.0

**Nextcloud version:** 15.0.0

**Updated from an older Nextcloud/ownCloud or fresh install:** updated from 14 (and 15 beta/rc)

**Where did you install Nextcloud from:** zip file

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.1.0
  - activity: 2.8.2
  - bookmarks: 0.15.1
  - calendar: 1.6.4
  - cloud_federation_api: 0.1.0
  - comments: 1.5.0
  - contacts: 3.0.1
  - dav: 1.8.0
  - deck: 0.5.2
  - encryption: 2.3.0
  - federatedfilesharing: 1.5.0
  - federation: 1.5.0
  - files: 1.10.0
  - files_external: 1.6.0
  - files_pdfviewer: 1.4.0
  - files_sharing: 1.7.0
  - files_texteditor: 2.7.0
  - files_trashbin: 1.5.0
  - files_versions: 1.8.0
  - files_videoplayer: 1.4.0
  - gallery: 18.2.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.3.0
  - mail: 0.11.0
  - news: 13.0.3
  - nextcloud_announcements: 1.4.0
  - notifications: 2.3.0
  - oauth2: 1.3.0
  - password_policy: 1.5.0
  - previewgenerator: 2.0.0
  - provisioning_api: 1.5.0
  - serverinfo: 1.5.0
  - sharebymail: 1.5.0
  - spreed: 5.0.0
  - support: 1.0.0
  - survey_client: 1.3.0
  - systemtags: 1.5.0
  - theming: 1.6.0
  - twofactor_backupcodes: 1.4.1
  - twofactor_totp: 2.1.0
  - updatenotification: 1.5.0
  - workflowengine: 1.5.0
Disabled:
  - admin_audit
  - bookmarks.bak
  - caniupdate
  - drawio
  - end_to_end_encryption
  - files_downloadactivity
  - firstrunwizard
  - user_external
  - user_ldap

```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            ""
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "15.0.0.10",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "theme": "",
        "maintenance": false,
        "maxZipInputSize": 0,
        "allowZipDownload": true,
        "secret": "***REMOVED SENSITIVE VALUE***",
        "app.mail.server-side-cache.enabled": true,
        "trashbin_retention_obligation": "auto",
        "preview_max_scale_factor": 1,
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "htaccess.RewriteBase": "\/",
        "overwrite.cli.url": "",
        "updater.release.channel": "stable",
        "mysql.utf8mb4": true,
        "mail_smtpmode": "sendmail",
        "mail_smtpauthtype": "LOGIN",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "loglevel": 2
    }
}

```
</details>

**Are you using external storage, if yes which one:**No

**Are you using encryption:** No, but the encryption app was enabled

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
[webdav] Fatal: Sabre\DAV\Exception: Error while copying file to target location (copied bytes: 0, expected filesize: 0 ) at <<closure>>

0. /var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php line 156
   put(null)
1. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 1096
   createFile("emptyfile", null)
2. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 525
   createFile("emptyfile", null, null)
3. <<closure>>
   httpPut(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
4. /var/www/nextcloud/3rdparty/sabre/event/lib/EventEmitterTrait.php line 105
   call_user_func_array([Sabre\DAV\CorePlugin {},"httpPut"], [Sabre\HTTP\Requ ... }])
5. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 479
   emit("method:PUT", [Sabre\HTTP\Requ ... }])
6. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 254
   invokeMethod(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
7. /var/www/nextcloud/apps/dav/appinfo/v1/webdav.php line 80
   exec()
8. /var/www/nextcloud/remote.php line 163
   require_once("/var/www/nextcl ... p")

PUT /remote.php/webdav/emptyfile
from 172.17.0.1 by admin at 2018-12-27T15:43:57+00:00


[PHP] Error: feof(): supplied resource is not a valid stream resource at /var/www/nextcloud/apps/dav/lib/Connector/Sabre/File.php#172

PUT /remote.php/webdav/emptyfile
from 172.17.0.1 by admin at 2018-12-27T15:43:57+00:00
```
</details>

I tried disabling all the apps I could and then re-enable them one for one. As soon as I re-enable the encryption app, the error occurs. I don't have encryption enabled though.
I tried on a fresh installation and enabling the encryption app (without configuring anything) gives the same error.

### PR DESCRIPTION
**Pull request by kesselb**
e60a829b42f0f4b74db835d8e10438a33e125051 local

```
Uncaught Error: Class 'OC\Log\ExceptionSerializer' not found in /var/www/html/lib/private/Log.php:316
Stack trace:
#0 /var/www/html/lib/private/Log/ErrorHandler.php(100): OC\Log->logException(Object(Error), Array)
#1 [internal function]: OC\Log\ErrorHandler::onAll(2, 'fwrite(): suppl...', '/var/www/html/a...', 62, Array)
#2 /var/www/html/apps-extra/files_antivirus/lib/Scanner/ExternalClam.php(62): fwrite(Resource id #1295, '\x00\x00\x00\x00')
#3 /var/www/html/apps-extra/files_antivirus/lib/Scanner/ScannerBase.php(143): OCA\Files_Antivirus\Scanner\ExternalClam->shutdownScanner()
#4 /var/www/html/apps-extra/files_antivirus/lib/AvirWrapper.php(118): OCA\Files_Antivirus\Scanner\ScannerBase->completeAsyncScan()
#5 [internal function]: OCA\Files_Antivirus\AvirWrapper->OCA\Files_Antivirus\{closure}()
#6 /var/www/html/apps/files_external/3rdparty/icewind/streams/src/CallbackWrapper.php(119): call_user_func(Object(Closure))
#7 [internal function]: Icewind\Streams\CallbackWrapper->stream_close()
#8 {main}
  thrown at /var/www/html/lib/private/Log.php#316
```

I run into the above error when scanning a chunked upload after assembly. files_antivirus is using a callback for [stream_close](https://github.com/nextcloud/files_antivirus/blob/f6c63f3a46550f040299c30bd7d7590741436c42/lib/AvirWrapper.php#L118) to fetch the status (infected yes/no) from clamav. When the stream is not closed explicitly it's happening so late that not even the autoloader is registered anymore. 

Fallback implementation for storage already closes the stream on close:
https://github.com/nextcloud/server/blob/dbf7905149222115a2cd0334efcf8c93afa8683e/lib/private/Files/Storage/Common.php#L873

ObjectStoreStorage is also closing the streams explicitly:
https://github.com/nextcloud/server/blob/dbf7905149222115a2cd0334efcf8c93afa8683e/lib/private/Files/ObjectStore/ObjectStoreStorage.php#L491
https://github.com/nextcloud/server/blob/dbf7905149222115a2cd0334efcf8c93afa8683e/lib/private/Files/ObjectStore/ObjectStoreStorage.php#L497

be3f4edf1f38b1ebfd91366334e5a3a91c63cffe encryption

1. The initial version https://github.com/nextcloud/server/commit/c6a48110bfbf5b3f100b0f2fc092e806e079e34e were closing the input stream. 
2. https://github.com/nextcloud/server/pull/13468 removed `fclose($stream)` to fix https://github.com/nextcloud/server/issues/13276. 
3. In the meantime https://github.com/nextcloud/server/pull/22739 update detection for failed writes to the storage. That's a better way to fix point 2. 

I'm able to upload empty files with and without encryption (via web and nautilus) to local storage and object store.

### CODE REVIEW

**Review by nickvergessen — APPROVED** [review]

**Comment by kesselb** [comment]
image

:worried:

**Comment by blizzz** [comment]
> image
> 
> worried

this is an abstract test class. For which implementation does it fail?

> image
> 
> worried

It is not closed in https://github.com/nextcloud/server/blob/d0383ddbae6eed17b6eb44e00dbe5912b52ab901/lib/private/Files/Storage/Wrapper/Encryption.php#L1032-L1037

File: lib/private/Files/Storage/Local.php [hunk 1]
```diff
@@ -568,6 +568,9 @@ public function moveFromStorage(IStorage $sourceStorage, $sourceInternalPath, $t
 
 	public function writeStream(string $path, $stream, int $size = null): int {
 		$result = $this->file_put_contents($path, $stream);
+		if (is_resource($stream)) {
+			fclose($stream);
+		}
 		if ($result === false) {
 			throw new GenericFileException("Failed write steam to $path");
```
Comment (acsfer) [inline lib/private/Files/Storage/Local.php:575]: Steam?

**Review by blizzz — APPROVED** [review]

**Pull request merged**